src/engine-actions/wait-for-swap-commitment.js

```diff
@@ -12,25 +12,50 @@ class CanceledSwapError extends Error {}
  */
 async function waitForSwapCommitment (swapHash) {
   return new Promise((resolve, reject) => {
+    // we use an array for timers since an invoice can be
+    // trigger the stream in an OPEN state multiple times.
+    const timers = []
+
     // subscribeSingleInvoice always sends out the initial invoice state
     const stream = subscribeSingleInvoice(swapHash, { client: this.client })
-    stream.on('end', () => reject(new Error(
-      'Stream ended while waiting for commitment on ' + swapHash)))
+    stream.on('end', () => {
+      stream.removeAllListeners()
+      reject(new Error('Stream ended while waiting for commitment on ' + swapHash))
+
+      // stop the timers on invoice expiration if they are still active
+      timers.forEach(clearTimeout)
+    })
     stream.on('data', function (invoice) {
       switch (invoice.state) {
         case INVOICE_STATES.OPEN:
+          const creationDate = new Date(invoice.creationDate * 1000)
+          const expirationDate = new Date(
+            creationDate.getTime() + (invoice.expiry * 1000)
+          )
+
+          // if the invoice is open but still expired, we consider it canceled
+          if (new Date() > expirationDate) {
+            reject(new CanceledSwapError(swapHash))
+            stream.cancel()
+            return
+          }
+          // once the invoice is expired, treat it as cancelled
```

Comment:
I'm not clear on the motivation for this change; naively it seems like it would be better to have different errors for cancellation vs expiration in case the caller wants to know.

Reply:
Perhaps, but how would you handle them differently? They are both cases where the invoice is not and will never be settled based on the action of this actor (either explicitly cancelling, or by implicitly cancelling through expiration).